nomos-da/network/core/src/swarm/common/monitor.rs

```diff
@@ -0,0 +1,387 @@
+        &self,
+        now: Instant,
+        window: Duration,
+        factor: U57F7,
+    ) -> U57F7 {
+        compute_failure_rate(
+            now,
+            self.last_dispersal_failure.unwrap_or(now),
+            self.dispersal_failures_rate,
+            window,
+            factor,
+        )
+    }
+
+    pub fn compute_sampling_failure_rate(
+        &self,
+        now: Instant,
+        window: Duration,
+        factor: U57F7,
+    ) -> U57F7 {
+        compute_failure_rate(
+            now,
+            self.last_sampling_failure.unwrap_or(now),
+            self.sampling_failures_rate,
+            window,
+            factor,
+        )
+    }
+
+    /// **Updates the replication failure rate**
+    pub fn compute_replication_failure_rate(
+        &self,
+        now: Instant,
+        window: Duration,
+        factor: U57F7,
+    ) -> U57F7 {
+        compute_failure_rate(
+            now,
+            self.last_replication_failure.unwrap_or(now),
+            self.replication_failures_rate,
+            window,
+            factor,
+        )
+    }
+}
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub struct DAConnectionMonitorSettings {
+    pub max_dispersal_failures: usize,
+    pub max_sampling_failures: usize,
+    pub max_replication_failures: usize,
+    pub malicious_threshold: usize,
+    pub failure_time_window: Duration,
+    pub time_decay_factor: U57F7,
+}
+
+pub struct DAConnectionMonitor {
+    peer_stats: HashMap<PeerId, PeerStats>,
+    settings: DAConnectionMonitorSettings,
+}
+
+impl DAConnectionMonitor {
+    pub fn new(settings: DAConnectionMonitorSettings) -> Self {
+        Self {
+            peer_stats: HashMap::new(),
+            settings,
+        }
+    }
+
+    fn evaluate_peer(&self, now: Instant, peer_id: &PeerId) -> PeerStatus {
+        if let Some(stats) = self.peer_stats.get(peer_id) {
+            // We need to recompute the failure rate upon the evaluation as time has moved on and
+            // the failure rate must have decayed..
+            let dispersal_rate = stats.compute_dispersal_failure_rate(
+                now,
+                self.settings.failure_time_window,
+                self.settings.time_decay_factor,
+            );
+            let sampling_rate = stats.compute_sampling_failure_rate(
+                now,
+                self.settings.failure_time_window,
+                self.settings.time_decay_factor,
+            );
+            let replication_rate = stats.compute_replication_failure_rate(
+                now,
+                self.settings.failure_time_window,
+                self.settings.time_decay_factor,
+            );
+
+            if dispersal_rate >= self.settings.malicious_threshold
+                || sampling_rate >= self.settings.malicious_threshold
+                || replication_rate >= self.settings.malicious_threshold
+            {
+                return PeerStatus::Malicious;
+            }
```

Comment:
I would extract this as a method. In fact, this should be part of the policies abstraction we talked about (?)

Reply:
Thats a good idea actually, the policy will be shared by the monitor and maintenance behaviours, and this implementation will provide stats to policy for evaluation. Will add a todo comment for now, implement connection policy in upcoming pr.